Propagator

jaescartin1 opened the pull request:
Decoupled timestep with a Propagator class common to all the tests.

This is the first step, generalize the loop. 

The second step, not implemented on this branch, will be add a list of conservedQuantities , maybe different between tests.

As Sebastian suggested, something like:

ParticlesData d = SedovDataGenerator...
// x, y, z, h, m are mandatory and assumed conserved
d.addConserved(&d.mui, &d.u, &d.vx, &d.vy, &d.vz, ...);
d.addSecondary(&d.ro, &d.p, &d.c, ...);

for (d.iteration = 0; d.iteration < maxStep; d.iteration++)
{
    propagator.hydroStep(domain, d);
    // file output if needed
}

but the second step (conserved and secondary quantities) should be develop in another PR

jaescartin1 committed:
Add propagator schema

jaescartin1 committed:
Add propagator to sedov.cpp and evrard.cpp

sekelle commented on include/tests/propagator.hpp:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    // Advance simulation by one step
+    void hydroStep()
```
Take the domain and ParticlesData here as arguments.

sekelle commented on include/tests/propagator.hpp:
```diff
@@ -0,0 +1,167 @@
+using namespace sphexa;
+using namespace sphexa::sph;
+
+template <class I, class T, class Accelerator = CpuTag>
```
These template arguments go to `hydroStep`, but as
`template<class DomainType, class ParticleDataType>`

sekelle commented on include/tests/propagator.hpp:
```diff
@@ -0,0 +1,167 @@
+
+    TaskList                   taskList;                        //
+
+    const bool                 gravity;                         //
```
You can remove anything related to gravity, we're going to make a separate class Propagator for gravity.
(There will be a lot of changes to how gravity is handled)

sekelle commented on include/tests/propagator.hpp:
```diff
@@ -0,0 +1,167 @@
+
+#include "../Timer.hpp"
+#include "../Task.hpp"
+#include "../ParticlesData.hpp"
```
remember to remove "../" once this file is in `include` (you can actually move it already now)

sekelle commented on include/tests/propagator.hpp:
```diff
@@ -0,0 +1,167 @@
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
```
The back-and-force between timing inside the propagator and outside it adds a lot of clutter.
Do you think you could move all the timing inside the propagator and not time the file I/O, such that
the file I/O only is part of the final runtime, but not the iteration steps?

sekelle commented:
Note on formatting: some changes are not compatible with the .clang-format in the top-level directory.
I'll address that if you don't have a formatting-tool at hand.

jaescartin1 committed:
Changes required by Sebastian

jaescartin1 commented:
Hi Sebastian,
       all the changes that you required were applied:
- hydrostep with domain and particleData as parameters
- template in function hydrostep, instead of in the class
- Removed gravity boolean and made another function in propagator with gravity
- Moved include/tests/propagator.hpp to include/ and renamed the includes inside
- Removed timer from propagator and included in app codes
- Now Public access to the parameter taskList, in order to reduce the access overhead

About formating, please, let me know the tool that give the format that you wish.

sekelle commented on include/propagator.hpp:
```diff
@@ -0,0 +1,184 @@
+
+    template <class DomainType, class ParticleDataType, class T>
+    void hydroStep(DomainType& domain, ParticleDataType& d)
+    {
```
```suggestion
    {
    using T = ParticleDataType::RealType;
```

sekelle commented on include/propagator.hpp:
```diff
@@ -0,0 +1,184 @@
+        return neighborsSum(taskList.tasks);
+    }
+
+    template <class DomainType, class ParticleDataType, class T>
```
```suggestion
    template <class DomainType, class ParticleDataType>
```

jaescartin1 committed:
Enabled timer inside the timestep loop and disable outside (IO)

sekelle commented on include/propagator.hpp:
```diff
@@ -0,0 +1,194 @@
+public:
+
+    TaskList           taskList;                                //
+    MasterProcessTimer timer;                                   //
```
can be private?

sekelle commented on src/evrard/evrard.cpp:
```diff
@@ -101,126 +105,120 @@ int main(int argc, char** argv)
-        timer.step("UpdateSmoothingLength");
-
-        size_t totalNeighbors = neighborsSum(taskList.tasks);
+        propagator.hydroStepGravity<DomainType, Dataset, Real>(domain, d);
```
```suggestion
        propagator.hydroStepGravity(domain, d);
```
Should be possible once the last template parameter is removed

jaescartin1 committed:
Update include/propagator.hpp

Co-authored-by: Sebastian Keller <44800242+sebkelle1@users.noreply.github.com>

jaescartin1 committed:
Update include/propagator.hpp

Co-authored-by: Sebastian Keller <44800242+sebkelle1@users.noreply.github.com>

jaescartin1 committed:
Update src/evrard/evrard.cpp

Co-authored-by: Sebastian Keller <44800242+sebkelle1@users.noreply.github.com>

jaescartin1 committed:
wrong compilation, dependent scope

jaescartin1 committed:
Fix one compilation error

sekelle commented:
cscs-ci run
